FeatureExtension Value At Risk, Expected Shortfall

damian-t opened the pull request:
Changes proposed in this pull request:
- Added class to keep track of the returns of a strategy (log and arithmetic)
- Added Value At Risk as an analysis criterion
- Added Expected Shortfall as an analysis criterion

- [x ] added an entry to the unreleased section of `CHANGES.md`

team172011 commented on ta4j-core/src/main/java/org/ta4j/core/analysis/Returns.java:
```diff
@@ -0,0 +1,176 @@
+            @Override
+            public Num calculate(Num x_new, Num x_old) {
+                // r_i = P_i/P_(i-1) - 1
+                return x_new.dividedBy(x_old).minus(x_new.numOf(1));
```
Performance would be better if you store `numOf(1)` as a class field instead of converting `int` to `num ` in every calculation

damian-t commented:
Included your remark of storing `numOf(1)` as a class variable. Also consistently compute the returns in Double precision (added another test for that)

team172011 commented:
Why are your specifying `Double` precision? You could just take the `numFunction` of the `TimeSeries`, so that the precision of the underlying TimeSeries will be guaranteed.

damian-t commented:
> Why are your specifying `Double` precision? You could just take the `numFunction` of the `TimeSeries`, so that the precision of the underlying TimeSeries will be guaranteed.

The `TimeSeries` object is only given in the constructor of `Returns` but the ReturnType is part of the `Return` class definition where it doesn't know about `TimeSeries`. Is there a way to overcome this?

team172011 commented:
>  Is there a way to overcome this?

At least for ARITHMETIC ReturnType you could write: 
`x_new.dividedBy(x_old).minus(one);`

And just specify for LOG the double precision

damian-t commented:
Thanks for the feedback, I updated the PR.
I also changed that the `Num` implementation of the LOG return is the same as for the corresponding timeseries, even though the computation is performed in `Double` precision.

damian-t commented:
@team172011 can this be merged?

milczarekIT commented on ta4j-core/src/main/java/org/ta4j/core/analysis/criteria/ValueAtRiskCriterion.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Override
+    public boolean betterThan(Num criterionValue1, Num criterionValue2) {
```
Greater ValueAtRisk is better? Hmm, I am not sure

damian-t replied:
Losses are negative, hence VaR is negative. Better means less negative, meaning "greater is better". I wrote a comment in `calculateVaR()`. I will also add a comment to the `betterThan()` function.

damian-t commented on ta4j-core/src/main/java/org/ta4j/core/analysis/Returns.java:
```diff
@@ -0,0 +1,184 @@
+ * </p>
+ * This class allows to compute the return rate of a price time-series
+ */
+public class Returns implements Indicator<Num> {
```
I tried to keep it consistent with the CashFlow class. That one implements Indicator<Num>

team172011 commented:
thanks!
